src/wallet.rs

```diff
@@ -0,0 +1,962 @@
+        };
+        Ok(addr)
+    }
+}
+
+impl TryFrom<String> for Address {
+    type Error = Error;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        let dec = bs58::decode(s).into_vec()?;
+        if dec.len() != SEED_SIZE {
+            return Err(Error::BadAddress);
+        }
+        let mut addr_bytes = [0u8; SEED_SIZE];
+        addr_bytes.copy_from_slice(&dec);
+        let addr = Address {
+            index: None,
+            psk: dusk_pki::PublicSpendKey::from_bytes(&addr_bytes)?,
+        };
+        Ok(addr)
+    }
+}
+
+impl TryFrom<&[u8; 64]> for Address {
+    type Error = Error;
+
+    fn try_from(bytes: &[u8; 64]) -> Result<Self, Self::Error> {
+        let addr = Address {
+            index: None,
+            psk: dusk_pki::PublicSpendKey::from_bytes(bytes)?,
+        };
+        Ok(addr)
+    }
+}
+
+impl PartialEq for Address {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index && self.psk == other.psk
+    }
+}
+
+impl std::hash::Hash for Address {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.index.hash(state);
+        self.psk.to_bytes().hash(state);
+    }
+}
+
+impl Display for Address {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", bs58::encode(self.psk.to_bytes()).into_string())
+    }
+}
+
+impl Debug for Address {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", bs58::encode(self.psk.to_bytes()).into_string())
+    }
+}
+
+/// Addresses holds address-related metadata that needs to be
+/// persisted in the wallet file.
+struct Addresses {
+    count: u64,
+}
+
+impl Default for Addresses {
+    fn default() -> Self {
+        Self { count: 1 }
+    }
+}
+
+impl Addresses {
+    pub(crate) fn from_bytes(buf: Vec<u8>) -> Result<Self, crate::Error> {
+        let mut u64_bytes = [0u8; 8];
+        u64_bytes.copy_from_slice(&buf);
+        Ok(Self {
+            count: u64::from_le_bytes(u64_bytes),
+        })
+    }
+
+    pub(crate) fn to_vec(&self) -> Vec<u8> {
+        self.count.to_le_bytes().to_vec()
+    }
+}
+
+/// Wrapper around `PathBuf` for wallet paths
+#[derive(PartialEq, Eq, Hash, Debug, Clone)]
+pub struct WalletPath(PathBuf);
+
+impl WalletPath {
+    /// Create a new wallet path from a directory and a name
+    pub fn new(dir: &Path, name: String) -> Self {
+        let mut pb = PathBuf::from(dir);
+        pb.push(name);
+        pb.set_extension("dat");
+        Self(pb)
+    }
+
+    /// Returns the filename of this path
+    pub fn name(&self) -> Option<String> {
+        // extract the name
+        let name = self.0.file_stem()?.to_str()?;
+        Some(String::from(name))
+    }
+
+    /// Returns current directory for this path
+    pub fn dir(&self) -> Option<PathBuf> {
+        let mut p = self.0.clone();
+        if p.pop() {
+            Some(p)
+        } else {
+            None
+        }
```

Comment:
You shouldn't `clone` and mutate, you should be able to do it as follow:
```suggestion
    let p = &self.0;
    p.parent().map(PathBuf::from)
```

Also, if you don't really need a `PathBuf` here, but just a `Path`, you could just have `p.parent()`.

Reply:
I didn't know about `parent()`! Great!